### ISSUE TITLE
RKE2 serving cert not rotating

### ISSUE THREAD

**Issue by QtSYnT** [rancher/rke2#4372]
**Environmental Info:**
RKE2 Version: 

rke2 version v1.24.14+rke2r1 (92fc41d4e8b155e83bfbca9feca91beaf06017a9)
go version go1.19.9 X:boringcrypto

Node(s) CPU architecture, OS, and Version: 

Linux rke2-cert-1 4.18.0-425.3.1.el8.x86_64 #1 SMP Fri Sep 30 11:45:06 EDT 2022 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:

Single node cluster

**Describe the bug:**

RKE2 rotates all certs on disk but the cert for the `rke2-serving` tls secret is not rotated thus eventually failing the cluster when it expires.

**Steps To Reproduce:**

This method of testing could be completely wrong so if you have a better way to handle this I'd love to give it a shot.

1. Single node RKE2 cluster is running
```
# Get current time
[root@rke2-cert-1 ~]# timedatectl | grep Local
               Local time: Thu 2023-06-15 12:49:45 EDT

# Get rke2-serving expire
[root@rke2-cert-1 ~]# kubectl get secret -n kube-system rke2-serving --template='{{index .data "tls.crt"}}' | base64 -d | openssl x509 -text | grep After
            Not After : Jun 14 16:08:29 2024 GMT
```
2. Set time to almost 1 year later
```
# Cert will expire in two weeks
[root@rke2-cert-1 ~]# timedatectl set-time "2024-06-01 12:49:45"
```

3. Restart rke2-server and check certs
```
[root@rke2-cert-1 ~]# systemctl restart rke2-server

[root@rke2-cert-1 ~]# timedatectl | grep Local
               Local time: Sat 2024-06-01 13:05:43 EDT

# rke2-serving is not rotated
[root@rke2-cert-1 ~]# kubectl get secret -n kube-system rke2-serving --template='{{index .data "tls.crt"}}' | base64 -d | openssl x509 -text | grep After
            Not After : Jun 14 16:08:29 2024 GMT

# Other certs were
[root@rke2-cert-1 ~]# openssl x509 -text -in /var/lib/rancher/rke2/server/tls/serving-kube-apiserver.crt | grep After
            Not After : Jun  1 16:51:20 2025 GMT

```

4. Set date to after expiration time for the original certs (but before rotated ones expire)
```
[root@rke2-cert-1 ~]# timedatectl set-time "2024-07-01 12:49:45"
```
5. Restart rke2-server
```
Jul 01 12:50:35 rke2-cert-1 rke2[52407]: time="2024-07-01T12:50:35-04:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": x509: certificate has expired or is not yet valid: current time 2024-07-01T12:50:35-04:00 is after 2024-06-14T16:08:29Z"
Jul 01 12:50:37 rke2-cert-1 rke2[52407]: time="2024-07-01T12:50:37-04:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": x509: certificate has expired or is not yet valid: current time 2024-07-01T12:50:37-04:00 is after 2024-06-14T16:08:29Z"
```
From this point on the rke2-server service will timeout and restart but never get past this error.

**Expected behavior:**

All certificates that are nearing their expiration date are rotated and the cluster will continue to function.

**Actual behavior:**

All certs on disk appear to rotate however the cert (in the datastore?) for rke2-serving is not rotated and thus the node fails to restart.

**Additional context / logs:**

[first-rke2-restart.log](https://github.com/rancher/rke2/files/11760864/first-rke2-restart.log)
[second-rke2-restart.log](https://github.com/rancher/rke2/files/11760866/second-rke2-restart.log)

**Comment by brandond** [comment]
The "second-rke2-restart" log shows the cert being updated:
```
Jul 01 12:50:33 rke2-cert-1 rke2[52407]: time="2024-07-01T12:50:33-04:00" level=info msg="certificate CN=rke2,O=rke2 will expire in -17.029214 days at 2024-06-14 16:08:29 +0000 UTC"
Jul 01 12:50:33 rke2-cert-1 rke2[52407]: time="2024-07-01T12:50:33-04:00" level=warning msg="unable to verify existing certificate: x509: certificate has expired or is not yet valid: current time 2024-07-01T12:50:33-04:00 is after 2024-06-14T16:08:29Z - signing operation may change certificate issuer"
Jul 01 12:50:33 rke2-cert-1 rke2[52407]: time="2024-07-01T12:50:33-04:00" level=info msg="certificate CN=rke2,O=rke2 signed by CN=rke2-server-ca@1686845309: notBefore=2023-06-15 16:08:29 +0000 UTC notAfter=2025-07-01 16:50:33 +0000 UTC"
```

It seems like dynamiclistener never sees the update though, probably because the secret controller startup is being blocked by agent startup, which in turn is stuck on the old cert.

**Comment by brandond** [comment]
You should be able to delete `/var/lib/rancher/rke2/server/tls/dynamic-cert.json` and then restart the service; can you confirm?

**Comment by QtSYnT** [comment]
> You should be able to delete `/var/lib/rancher/rke2/server/tls/dynamic-cert.json` and then restart the service to unblock this; can you confirm?

So that got kubelet to start but now it hangs later on in startup:

```
Jul 01 12:51:03 rke2-cert-1 rke2[25634]: time="2024-07-01T12:51:03-04:00" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": x509: certificate has expired or is not yet valid: current time 2024-07-01T12:51:03-04:00 is after 2024-06-14T16:08:29Z"
Jul 01 12:51:08 rke2-cert-1 rke2[25634]: time="2024-07-01T12:51:08-04:00" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": x509: certificate has expired or is not yet valid: current time 2024-07-01T12:51:08-04:00 is after 2024-06-14T16:08:29Z"
```

Starting from step 5 I did the following to get to here:

```
/usr/local/bin/rke2-killall.sh
rm -f /var/lib/rancher/rke2/server/tls/dynamic-cert.json
systemctl restart rke2-server
```

Full log:
[rke2-server.log](https://github.com/rancher/rke2/files/11761832/rke2-server.log)

**Comment by brandond** [comment]
Is the apiserver static pod still running? You didn't mention running the killall script, so I suspect it is. Can you also run `kubectl delete secret -n kube-system rke2-serving`? Do that alongside removing the dynamic-cert.json from disk, and then restart the service.

**Comment by brandond** [comment]
I am also curious if this is still a problem if you run the killall script before setting the system time forward. Just setting the time forward and then restarting the service is not a realistic test, and there are bits of etcd and kubernetes that will not react well to drastic time changes. I suspect (although I could be wrong) that this is handled better if RKE2 is completely terminated via the killall script when you set the time forward.

**Comment by QtSYnT** [comment]
Yeah, I that was one of my concerns about this test so I was also looking to see how you guys tested this in the past. 

I'll give it a shot and will report back how it goes.

**Comment by QtSYnT** [comment]
The following did work and the cert was rotated:
1. Start RKE2 cluster
2. Set time within 90 window
3. Restart rke2-server service
4. Check certs on disk are rotated but rke2-serving is not
5. Set time past original cert expire but before rotated cert expiration
6. Restart rke2-server service
7. RKE2 fails to start: `Jul 01 17:22:53 rke2-cert-1 rke2[7608]: time="2024-07-01T17:22:53-04:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": x509: certificate has expired or is not yet valid: current time 2024-07-01T17:22:53-04:00 is after 2024-06-14T21:10:31Z"`
8. Delete rke2-serving secret and dynamic-cert.json
9. Restart rke2-server service
10. RKE2 comes back up and dumping rke2-serving cert shows a rotated cert

The following did not work (kill rke2 and all pods before changing the time):

1. Start RKE2 cluster
2. Kill cluster and ensure no pods are running
3. Set time within the 90 renewal window
4. Certs are renewed but rke2-serving hasn't refreshed
5. Kill cluster and ensure no pods are running
6. Set time past the original cert expiration but not past rotated certs
7. Start rke2-server
8. Get `Jul 01 17:26:20 rke2-cert-1 rke2[9360]: time="2024-07-01T17:26:20-04:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": x509: certificate has expired or is not yet valid: current time 2024-07-01T17:26:20-04:00 is after 2024-06-14T21:10:31Z"`

**Comment by brandond** [comment]
> 2. Kill cluster and ensure no pods are running
> 3. Set time within the 90 renewal window
> 4. Start rke2-server

Can you check the logs to see if you see the cert being signed? Even in the logs you shared earlier, we saw it being updated on startup when expired.

We haven't touched any of this logic in a bit so I'm not sure why this would have regressed. We did just bump the version of dynamiclistener in this month's releases, I wonder if you get similar results from [v1.24.15-rc1+rke2r1](https://github.com/rancher/rke2/releases/tag/v1.24.15-rc1%2Brke2r1)

**Comment by QtSYnT** [comment]
For some background, we ran into the same issue on 1.24.4 with rke2-serving not renewing so I pulled in 1.24.14 to do more thorough testing for the purpose of this issue.

Anyways, here are the log outputs for step 4 and step 8 respectively:

[exec-1.log](https://github.com/rancher/rke2/files/11764026/exec-1.log)
[exec-2.log](https://github.com/rancher/rke2/files/11764027/exec-2.log)

I'll pull in the release candidate tomorrow and see if that fixes things.

**Comment by QtSYnT** [comment]
Alright, 1.24.15-rc1 seems to work as expected:

```
# Check kube api cert on disk
[root@rke2-cert-1 ~]# openssl x509 -text -in /var/lib/rancher/rke2/server/tls/serving-kube-apiserver.crt | grep After
            Not After : Jun 14 21:28:51 2024 GMT

# Check rke2-serving tls secret
[root@rke2-cert-1 ~]# kubectl get secret -n kube-system rke2-serving --template='{{index .data "tls.crt"}}' | base64 -d | openssl x509 -text | grep After
            Not After : Jun 14 21:28:52 2024 GMT

# Set time within 90 day renewal window
[root@rke2-cert-1 ~]# timedatectl set-time "2024-06-01 17:34:09"

# Restart rke2
[root@rke2-cert-1 ~]# systemctl restart rke2-server

# Check kube api cert on disk
[root@rke2-cert-1 ~]# openssl x509 -text -in /var/lib/rancher/rke2/server/tls/serving-kube-apiserver.crt | grep After
            Not After : Jun  1 21:35:34 2025 GMT

# Check rke2-serving tls secret
[root@rke2-cert-1 ~]# kubectl get secret -n kube-system rke2-serving --template='{{index .data "tls.crt"}}' | base64 -d | openssl x509 -text | grep After
            Not After : Jun  1 21:37:16 2025 GMT
```
Repeated a few times to ensure the certs continue to renew for multiple years.

**Comment by brandond** [comment]
OK. I was pretty sure that we'd fixed that on the dynamiclistener side a while back, I think we just missed pulling that through here into RKE2.

**Comment by marvinlnnx** [comment]
it is Oct 2024, still this problem persist, does Rancher team wants to implement a permanent fix for this problem ?
I check the date on my server is OK, I have used Rancher UI and installed more than 10 clusters, all the same configuration, 2 of them only got bootstrapped success  the rest all failed due to **certificate issue**  😡

I rotate the certs and then I checked the certs using `rke2 certificate rotate` and `rke2 certificate check`

```
rke2 certificate check
INFO[0000] Server detected, checking agent and server certificates
INFO[0000] Checking certificates for admin
INFO[0000] Checking certificates for controller-manager
INFO[0000] Checking certificates for scheduler
INFO[0000] Checking certificates for supervisor
INFO[0000] Checking certificates for kube-proxy
INFO[0000] Checking certificates for kubelet
INFO[0000] Checking certificates for rke2-controller
INFO[0000] Checking certificates for api-server
INFO[0000] Checking certificates for auth-proxy
INFO[0000] Checking certificates for cloud-controller
INFO[0000] Checking certificates for etcd
```

and then I started the `rke2-server` as soon as the server start the cers become invalid again 

```
rke2 certificate check
INFO[0000] Server detected, checking agent and server certificates
INFO[0000] Checking certificates for supervisor
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-supervisor.crt: certificate CN=system:rke2-supervisor,O=system:masters is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-supervisor.crt: certificate CN=rke2-client-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
INFO[0000] Checking certificates for kube-proxy
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-kube-proxy.crt: certificate CN=system:kube-proxy is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-kube-proxy.crt: certificate CN=rke2-client-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
INFO[0000] Checking certificates for kubelet
INFO[0000] Checking certificates for admin
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-admin.crt: certificate CN=system:admin,O=system:masters is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-admin.crt: certificate CN=rke2-client-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
INFO[0000] Checking certificates for cloud-controller
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-rke2-cloud-controller.crt: certificate CN=rke2-cloud-controller-manager is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-rke2-cloud-controller.crt: certificate CN=rke2-client-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
INFO[0000] Checking certificates for scheduler
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-scheduler.crt: certificate CN=system:kube-scheduler is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-scheduler.crt: certificate CN=rke2-client-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
INFO[0000] Checking certificates for etcd
ERRO[0000] /var/lib/rancher/rke2/server/tls/etcd/client.crt: certificate CN=etcd-client is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/etcd/client.crt: certificate CN=etcd-server-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/etcd/server-client.crt: certificate CN=etcd-server is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/etcd/server-client.crt: certificate CN=etcd-server-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/etcd/peer-server-client.crt: certificate CN=etcd-peer is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/etcd/peer-server-client.crt: certificate CN=etcd-peer-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
INFO[0000] Checking certificates for rke2-controller
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-rke2-controller.crt: certificate CN=system:rke2-controller is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-rke2-controller.crt: certificate CN=rke2-client-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
INFO[0000] Checking certificates for api-server
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-kube-apiserver.crt: certificate CN=system:apiserver,O=system:masters is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-kube-apiserver.crt: certificate CN=rke2-client-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/serving-kube-apiserver.crt: certificate CN=kube-apiserver is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/serving-kube-apiserver.crt: certificate CN=rke2-server-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
INFO[0000] Checking certificates for auth-proxy
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-auth-proxy.crt: certificate CN=system:auth-proxy is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-auth-proxy.crt: certificate CN=rke2-request-header-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
INFO[0000] Checking certificates for controller-manager
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-controller.crt: certificate CN=system:kube-controller-manager is not valid before 2024-10-17T04:01:49Z
ERRO[0000] /var/lib/rancher/rke2/server/tls/client-controller.crt: certificate CN=rke2-client-ca@1729137709 is not valid before 2024-10-17T04:01:49Z
```




and here is some errors:

```
Oct 17 06:21:51 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:21:51+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:21:51+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:21:53 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:21:53+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:21:53+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:21:55 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:21:55+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:21:55+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:21:57 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:21:57+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:21:57+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:21:59 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:21:59+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:21:59+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:01 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:01+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:01+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:03 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:03+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:03+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:05 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:05+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:05+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:07 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:07+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:07+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:09 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:09+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:09+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:11 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:11+03:00" level=info msg="Waiting for cri connection: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/k3s/containerd/containerd.sock: connect: connection refused\""
Oct 17 06:22:11 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:11+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:11+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:13 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:13+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:13+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:15 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:15+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:15+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:17 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:17+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:17+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:19 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:19+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:19+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:20 ocl-stg-cp-wsknh-thxgd rke2[5252]: {"level":"warn","ts":"2024-10-17T06:22:20.629612+0300","logger":"etcd-client","caller":"v3@v3.5.13-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0016d05a0/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: last connection error: connection error: desc = \"transport: Error while dialing: dial tcp 127.0.0.1:2379: connect: connection refused\""}
Oct 17 06:22:20 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:20+03:00" level=info msg="Failed to test data store connection: context deadline exceeded"
Oct 17 06:22:21 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:21+03:00" level=info msg="Waiting for etcd server to become available"
Oct 17 06:22:21 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:21+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:21+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:23 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:23+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:23+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:25 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:25+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:25+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:27 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:27+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:27+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:29 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:29+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:29+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:31 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:31+03:00" level=info msg="Waiting for cri connection: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/k3s/containerd/containerd.sock: connect: connection refused\""
Oct 17 06:22:31 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:31+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:31+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:33 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:33+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:33+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:35 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:35+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:35+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:37 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:37+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:37+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:39 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:39+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:39+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:41 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:41+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:41+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:43 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:43+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:43+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:45 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:45+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:45+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:47 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:47+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:47+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:49 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:49+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:49+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:51 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:51+03:00" level=info msg="Waiting for etcd server to become available"
Oct 17 06:22:51 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:51+03:00" level=info msg="Waiting for cri connection: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/k3s/containerd/containerd.sock: connect: connection refused\""
Oct 17 06:22:51 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:51+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:51+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:53 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:53+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:53+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:55 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:55+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:55+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:55 ocl-stg-cp-wsknh-thxgd rke2[5252]: {"level":"warn","ts":"2024-10-17T06:22:55.629941+0300","logger":"etcd-client","caller":"v3@v3.5.13-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0016d05a0/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: last connection error: connection error: desc = \"transport: Error while dialing: dial tcp 127.0.0.1:2379: connect: connection refused\""}
Oct 17 06:22:55 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:55+03:00" level=info msg="Failed to test data store connection: context deadline exceeded"
Oct 17 06:22:57 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:57+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:57+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:22:59 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:22:59+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:22:59+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:23:01 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:23:01+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:23:01+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:23:03 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:23:03+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:23:03+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:23:05 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:23:05+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:23:05+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:23:07 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:23:07+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:23:07+03:00 is before 2024-10-17T04:01:49Z"
Oct 17 06:23:09 ocl-stg-cp-wsknh-thxgd rke2[5252]: time="2024-10-17T06:23:09+03:00" level=error msg="CA cert validation failed: Get \"https://127.0.0.1:9345/cacerts\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-10-17T06:23:09+03:00 is before 2024-10-17T04:01:49Z"
```

**Comment by marvinlnnx** [comment]
this works for me if someone in 2024 having cert issue:

```
systemctl stop rke2-server
rm -rf /var/lib/rancher/rke2/server/tls/*
rm -rf /var/lib/rancher/rke2/server/db/etcd/member/*
rke2-killall.sh
rke2 certificate rotate
systemctl start rke2-server
```

I hope rancher will provide an easier way to manage the certs, for instance increasing expiry and fixing this errors.

**Comment by brandond** [comment]
> `rm -rf /var/lib/rancher/rke2/server/db/etcd/member/*`

This is bad advice. Don't do that. Don't delete files out from under from your etcd datastore!

This issue was closed by the reporter almost a year and a half ago and is as far as I know no longer an issue with recent releases of RKE2. If you're still experiencing problems, please open a new issue describing specifically what you're seeing, or at least add a comment here with useful steps to reproduce your problem.
